docs/spec/overload.rst

```diff
@@ -99,8 +109,397 @@ variable is not sufficient.
 
 Another important difference between type variables such as ``AnyStr``
 and using ``@overload`` is that the prior can also be used to define
-constraints for generic class type parameters.  For example, the type
+constraints for generic class type parameters. For example, the type
 parameter of the generic class ``typing.IO`` is constrained (only
 ``IO[str]``, ``IO[bytes]`` and ``IO[Any]`` are valid)::
 
   class IO(Generic[AnyStr]): ...
+
+
+Invalid overload definitions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Type checkers should enforce the following rules for overload definitions.
+
+At least two ``@overload``-decorated definitions must be present. If only
+one is present, an error should be reported.
+
+The ``@overload``-decorated definitions must be followed by an overload
+implementation, which does not include an ``@overload`` decorator. Type
+checkers should report an error or warning if an implementation is missing.
+Overload definitions within stub files, protocols, and abstract base classes
+are exempt from this check.
+
+If an overload is decorated with ``@staticmethod`` or ``@classmethod``,
+all overloads must be similarly decorated. The implementation,
+if present, must be decorated in the same manner. Type checkers should report
+an error if these conditions are not met.
+
+If one or more overloads are decorated with ``@final`` or ``@override`` but the
+implementation is not, an error should be reported.
```

Comment:
The idea behind this rule is that the implementation normally determines whether the function is `@final` or `@override`. However, there are cases where it's legitimate for an implementation to be omitted (e.g. in stubs or protocol definitions). In those cases, the `@final` or `@override` decorator needs to be applied to the overloads.

I agree that it would be simplest to enforce consistency. However, this isn't what type checkers do today. Mypy implements the rule as stated, and pyright copied it. My concern about changing the rule in the spec is that it will cause a lot of churn in existing code bases. Maybe the simplification is worth it? I could go either way here. 

@carljm, what are your thoughts now that you have this additional context?

Reply:
I'm not intending to advocate for any change to the rule here, I'm just trying to understand what the rule is; I don't think the current text is sufficiently clear.

I think I'm gathering from your reply that the current text should be understood as a narrow rule determining "under what circumstances should a type checker issue an error specifically regarding improper use of `@final` or `@override` on overloads?" But that's not the only question that must be answered here. If the rule is anything more complex than "all or nothing," then this text also needs to clarify under what circumstances a type checker should consider the overloaded method as actually decorated with `@final` or `@override`.

These are the questions I asked in my original comment:

> Is it is permissible to decorate the implementation with @final or @override, but not the overloads? Is it permissible to decorate some overloads, but not others? If any of these are permissible, what is the meaning of this mixed decoration?

I think I gather from your reply that these are both permissible, because the current text does not specifically prohibit them. And I think I also gather that if there is an implementation present, the presence or absence of the decorator on the implementation is all that really matters. I still don't know what the interpretation should be if some overloads are decorated and others are not, and there is no implementation.

Reply:
Yes, I see your point. I took another shot at this. It turns out that mypy's and pyright's behaviors differ slightly here. I think that both of them are reasonable and defensible, but we should converge on one. I've opted to specify mypy's behavior since its implementation predates pyright's.